### ISSUE TITLE
Scroll element into view before click

### ISSUE THREAD

**Issue by dbtfsbre** [Roemer/FlaUI#111]
It seems that currently there is no automated way to scroll an element into view before clicking onto it.
FlaUI.Core.AutomationElements.Infrastructure.AutomationElement.PerformMouseAction(bool, Action) calls GetClickablePoint() which throws an exception if the control is out of view, at least in my case with a WinForms application.

Does UIA provide any methods for this?
If not, should we search for elements which support the ScrollPattern among the Parents of the control to be clicked and check if their BoundingRectangle contains the control and if not, scroll them?

**Comment by Roemer** [comment]
There is a `ScrollIntoView` in the `ScrollItemPattern`. So items that support that pattern can be scrolled to with that.

**Comment by dbtfsbre** [comment]
Unfortunately in my case the control which is out of view does not support the ScrollItemPattern, although its ancestor supports the ScrollPattern. The only way of getting around (except changing the tested application which is very complex) would be to implement something which finds the scrollable ancestor(s) and scroll.

**Comment by maxinfet** [comment]
@dbtfsbre I have had to do this in our application as well since our grids do not support ScrollItemPattern. Essentially I made it so when retrieving the a row I would wrap it in a decorator that had a callback on it. Then I could call BringIntoView on the row and it would tell the grid to bring itself into view.

**Comment by dbtfsbre** [comment]
Just wanted to bring this up to ask if this kind of logic should be implemented in FlaUI.
Now I have implemented the described logic and it is working quite well.

@maxinfet In my case it is not an option to change the code of the tested application.

**Comment by dbtfsbre** [comment]
Also, if this would be added to FlaUI, should it be called by default before each MouseOperation on an AutomationElement? The method would go through all ancestors to verify if any of them needs to be scrolled (it is possible that multiple views need to be scrolled).
I didn't see any performance influence with WinForms and WPF controls, but still, the whole tree would be walked through in every call.

**Comment by danhi** [comment]
I had a similar problem, and here is what I did.  Count is the number of rows (in my case in a listbox).  As I process them, I check to see if the IsOffscreen is true, and if it is then I use the ScrollDown method of the vertical scrollbar.
            ' make sure the row is visible (or else the capture doesn't work)
For i As Integer = 0 To count - 1
            If i < count - 1 AndAlso uie.lv.Rows(i + 1).IsOffscreen Then ' need to look at the next one, because if it's partially onscreen it returns false
                ' need to scroll down
                If scrollBar Is Nothing Then
                    LogError("No scroll bar exists when an element is offscreen.")
                Else
                    scrollBar.ScrollDown()
                End If
            End If

**Comment by dbtfsbre** [comment]
@maxinfet Yes, it would be nice to see your solution. So far I am using the code in my pull request because in complex views with 3rd party controls this is reliably working. However it might be better to use the scrollbar instead of only the ScrollPattern of parent views.

**Comment by dbtfsbre** [comment]
In my case the problem is an application window with different docking areas. In this case, if one of the docked views has to be scrolled, the controls within that view have been rendered already.

In case of a list view or grid this might not be the case, so it is unknown where the control actually is before scrolling. This requires scrolling for an unknown number of times and if the ScrollPattern is not available on parent controls, also for an unknown value.

**Comment by dbtfsbre** [comment]
Created new pull request #160 which only scrolls hidden standard controls into view using the ScrollPattern
